common/splice_script.c

```diff
@@ -0,0 +1,2696 @@
+#include "config.h"
+#include <assert.h>
+#include <bitcoin/base58.h>
+#include <bitcoin/pubkey.h>
+#include <ccan/mem/mem.h>
+#include <ccan/str/hex/hex.h>
+#include <ccan/tal/str/str.h>
+#include <common/amount.h>
+#include <common/bech32.h>
+#include <common/json_parse.h>
+#include <common/json_parse_simple.h>
+#include <common/splice_script.h>
+#include <ctype.h>
+#include <inttypes.h>
+
+#define ARROW_SYMBOL "->"
+#define PIPE_SYMBOL '|'
+#define AT_SYMBOL '@'
+#define COLON_SYMBOL ':'
+#define Q_SYMBOL '?'
+#define WILD_SYMBOL '*'
+#define PLUS_SYMBOL '+'
+#define MINUS_SYMBOL '-'
+#define FEE_SYMBOL "fee"
+
+#define PERCENT_REGEX "^([0-9]*)[.]?([0-9]*)%$"
+#define Q_REGEX "^\\?$"
+#define WILD_REGEX "^\\*$"
+#define CHANID_REGEX "^[0-9A-Za-z]{64}$"
+#define NODEID_REGEX "^0[23][0-9A-Za-z]{64}$"
+#define WALLET_REGEX "^wallet$"
+#define FEE_REGEX "^" FEE_SYMBOL "$"
+#define SATM_REGEX "^([0-9]*)[.]?([0-9]*)[Mm]$"
+#define SATK_REGEX "^([0-9]*)[.]?([0-9]*)[Kk]$"
+
+#define CODE_SNIPPET_PADDING 80
+
+/* Minimum # of matching charaters to autocomplete nodeid or chanid */
+#define NODEID_MIN_CHARS 4
+#define CHANID_MIN_CHARS 4
+
+/* Token types from simplest to most complex. */
+enum token_type {
+	TOK_CHAR,
+	TOK_DELIMITER, /* Newline or semicolon. Guaranteed one at end. */
+	TOK_ARROW,
+	TOK_STR,
+	TOK_PIPE, /* lease separator "|", ex 5M|3M (add 5M, lease 3M) */
+	TOK_COLON, /* node query separator ":", ex nodeid:0 */
+	TOK_ATSYM, /* lease rate separator "@", ex 3M@2% */
+	TOK_PLUS,
+	TOK_MINUS,
+	TOK_PERCENT, /* ie "80%" */
+	TOK_SATS, /* ie 8M or 0 */
+	TOK_QUESTION, /* ie "?" */
+	TOK_WILDCARD, /* ie "*" */
+	TOK_FEE, /* ie the word "fee" */
+	TOK_CHANID,
+	TOK_WALLET,
+	TOK_FEERATE, /* The fee rate */
+	TOK_NODEID,
+	TOK_BTCADDR,
+	TOK_CHANQUERY, /* ie nodeid:? */
+	TOK_MULTI_CHANID, /* Matches stored in ->right */
+	TOK_LEASEREQ,
+	TOK_LEASERATE, /* ie @2% */
+	TOK_SEGMENT, /* An entire line or semicolon separated segment */
+};
+
+#define TOKEN_FLAG_FEERATE_NEGATIVE 0x01
+
+struct token {
+	enum token_type type;
+	size_t script_index; /* For error messages */
+	char c;
+	char *str, *rstr;
+	u32 ppm;
+	struct amount_sat amount_sat;
+	struct node_id *node_id;
+	struct channel_id *chan_id;
+	struct token *left, *middle, *right;
+	u32 flags;
+};
+
+static struct token *new_token(const tal_t *ctx, enum token_type token_type,
+			       size_t script_index)
+{
+	struct token *token = talz(ctx, struct token);
```

Comment:
I prefer doing it manually because each of these `new_token_*` functions would only be used once. I also like when the code shows `struct->param = value` so `param` is visible as opposed to hidden by the function call -- which is extra useful when doing tokens where `param`s have a tendency to get overloaded.

Reply:
Got rid of the talz usage and explicitly set values to 0